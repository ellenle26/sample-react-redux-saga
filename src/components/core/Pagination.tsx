import classNames from "classnames";
import { memo } from "react";

const Pagination: React.FC<PaginationProps> = (props) => {
    const {activedPage, total, paginationItems, onPageChange} = props;
    return <div className="flex flex-row justify-center" role={"pagination"}>
        <div className={classNames("flex items-center justify-center w-10 h-10 min-w-10 min-h-10 cursor-pointer", {
            "cursor-pointer": activedPage > 0
        })} onClick={() => activedPage > 0  && onPageChange && onPageChange(activedPage - 1)}>
            {activedPage > 0 && <img src="/icons/left_icon.svg"/>}
        </div>
        {paginationItems.map(value => <div key={value} className = {classNames("flex text-primary cursor-pointer font-medium text-base items-center justify-center px-2",{
            "bg-secondary rounded-full cursor-default w-10 h-10 min-w-10 min-h-10 p-0": activedPage + 1 === value
        })} onClick = {() => {
            if (value !== "...") {
                onPageChange && onPageChange(value - 1);
            } else {
                const index = paginationItems.indexOf("...");
                onPageChange && onPageChange(paginationItems[index - 1]);
            }
        }}>{value}</div>)}
        <div className={classNames("flex items-center justify-center w-10 h-10 min-w-10 min-h-10", {
            "cursor-pointer": activedPage < total - 1 
        })} onClick={() => activedPage < total - 1  && onPageChange && onPageChange(activedPage + 1)}>
            {activedPage < total - 1 && <img src="/icons/right_icon.svg"/>}
        </div>
    </div>
}

interface PaginationProps {
    total: number;
    activedPage: number;
    paginationItems: any[];
    onPageChange?: (value: number) => void;
}

export default memo(Pagination);