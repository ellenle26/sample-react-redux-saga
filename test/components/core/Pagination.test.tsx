/**
 * @jest-environment jsdom
*/

import Pagination from "@components/core/Pagination";
import React from "react";

import { fireEvent,render } from "../../test-utils";


test("snapshot testing", () => {
  const paginationItems = [1, 2, "...", 6, 7];
  const renderResult = render(<Pagination total={7} activedPage ={0} paginationItems = {paginationItems} />);
  expect(renderResult.asFragment()).toMatchSnapshot();
});


test("left arrow cannot be clicked testing", () => {
  const handleClick = jest.fn();
  const paginationItems = [1, 2, "...", 6, 7];
  const renderResult = render(<Pagination total={7} activedPage ={0} paginationItems = {paginationItems} onPageChange = {handleClick} />);
  expect(renderResult.container.children.length).toEqual(1);
  expect(renderResult.container.children[0].children.length).toEqual(7);
  fireEvent.click(renderResult.container.children[0].children[0]);
  expect(handleClick).toHaveBeenCalledTimes(0);
});

test("click on normal item testing", () => {
  const handleClick = jest.fn();
  const paginationItems = [1, 2, "...", 6, 7];
  const renderResult = render(<Pagination total={7} activedPage ={0} paginationItems = {paginationItems} onPageChange = {handleClick} />);
  expect(renderResult.container.children.length).toEqual(1);
  expect(renderResult.container.children[0].children.length).toEqual(7);
  fireEvent.click(renderResult.container.children[0].children[1]);
  expect(handleClick).toHaveBeenCalledTimes(1);
});

test("click on ... item testing", () => {
  const handleClick = jest.fn();
  const paginationItems = [1, 2, "...", 6, 7];
  const renderResult = render(<Pagination total={7} activedPage ={0} paginationItems = {paginationItems} onPageChange = {handleClick} />);
  expect(renderResult.container.children.length).toEqual(1);
  expect(renderResult.container.children[0].children.length).toEqual(7);
  fireEvent.click(renderResult.container.children[0].children[3]);
  expect(handleClick).toHaveBeenCalledTimes(1);
});

test("right arrow cannot be clicked testing", () => {
  const handleClick = jest.fn();
  const paginationItems = [1, 2, "...", 6, 7];
  const renderResult = render(<Pagination total={7} activedPage ={6} paginationItems = {paginationItems} onPageChange = {handleClick} />);
  expect(renderResult.container.children.length).toEqual(1);
  expect(renderResult.container.children[0].children.length).toEqual(7);
  fireEvent.click(renderResult.container.children[0].children[6]);
  expect(handleClick).toHaveBeenCalledTimes(0);
});

test("left arrow cannot be clicked testing", () => {
  const handleClick = jest.fn();
  const paginationItems = [1, 2, "...", 6, 7];
  const renderResult = render(<Pagination total={7} activedPage ={1} paginationItems = {paginationItems} onPageChange = {handleClick} />);
  expect(renderResult.container.children.length).toEqual(1);
  expect(renderResult.container.children[0].children.length).toEqual(7);
  fireEvent.click(renderResult.container.children[0].children[0]);
  expect(handleClick).toHaveBeenCalledTimes(1);
});

test("right arrow can be clicked testing", () => {
  const handleClick = jest.fn();
  const paginationItems = [1, 2, "...", 6, 7];
  const renderResult = render(<Pagination total={7} activedPage ={5} paginationItems = {paginationItems} onPageChange = {handleClick} />);
  expect(renderResult.container.children.length).toEqual(1);
  expect(renderResult.container.children[0].children.length).toEqual(7);
  fireEvent.click(renderResult.container.children[0].children[6]);
  expect(handleClick).toHaveBeenCalledTimes(1);
});