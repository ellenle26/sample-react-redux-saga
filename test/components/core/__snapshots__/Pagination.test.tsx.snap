// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row justify-center"
    role="pagination"
  >
    <div
      class="flex items-center justify-center w-10 h-10 min-w-10 min-h-10 cursor-pointer"
    />
    <div
      class="flex text-primary cursor-pointer font-medium text-base items-center justify-center px-2 bg-secondary rounded-full cursor-default w-10 h-10 min-w-10 min-h-10 p-0"
    >
      1
    </div>
    <div
      class="flex text-primary cursor-pointer font-medium text-base items-center justify-center px-2"
    >
      2
    </div>
    <div
      class="flex text-primary cursor-pointer font-medium text-base items-center justify-center px-2"
    >
      ...
    </div>
    <div
      class="flex text-primary cursor-pointer font-medium text-base items-center justify-center px-2"
    >
      6
    </div>
    <div
      class="flex text-primary cursor-pointer font-medium text-base items-center justify-center px-2"
    >
      7
    </div>
    <div
      class="flex items-center justify-center w-10 h-10 min-w-10 min-h-10 cursor-pointer"
    >
      <img
        src="/icons/right_icon.svg"
      />
    </div>
  </div>
</DocumentFragment>
`;
